source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+#include <dem/find_cell_neighbors.h>
+#include <dem/insertion_file.h>
+#include <dem/insertion_list.h>
+#include <dem/ray_tracing.h>
+#include <dem/read_mesh.h>
+
+#include <sys/stat.h>
+
+
+template <int dim>
+RayTracingSolver<dim>::RayTracingSolver(
+  RayTracingSolverParameters<dim> parameters,
+  DEMSolverParameters<dim>        dem_parameters)
+  : mpi_communicator(MPI_COMM_WORLD)
+  , n_mpi_processes(Utilities::MPI::n_mpi_processes(mpi_communicator))
+  , this_mpi_process(Utilities::MPI::this_mpi_process(mpi_communicator))
+  , pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+  , parameters(parameters)
+  , dem_parameters(dem_parameters)
+  , triangulation(this->mpi_communicator)
+  , mapping(1)
+  , particle_handler(triangulation,
+                     mapping,
+                     DEMProperties::PropertiesIndex::n_properties)
+  , photon_handler(triangulation, mapping, dim)
+  , computing_timer(this->mpi_communicator,
+                    this->pcout,
+                    TimerOutput::summary,
+                    TimerOutput::wall_times)
+  , background_dh(triangulation)
+  , photon_displacement_vector(
+      parameters.ray_tracing_info.photon_displacement_vector)
+{
+  if (parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic_with_sparse_contacts ||
+      parameters.model_parameters.load_balance_method ==
+        Parameters::Lagrangian::ModelParameters<
+          dim>::LoadBalanceMethod::dynamic)
+    throw std::runtime_error(
+      "The \"dynamic\" and \"dynamic_with_sparse_contacts\""
+      "load balancing methods are not supported");
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::setup_parameters()
+{
+  // Print simulation starting information
+  pcout << std::endl;
+  std::stringstream ss;
+  ss << "Running on " << n_mpi_processes << " rank(s)";
+  announce_string(pcout, ss.str(), '*');
+
+  // Check if the output directory exists
+  std::string output_dir_name = parameters.simulation_control.output_folder;
+  struct stat buffer;
+
+  // If output directory does not exist, create it
+  if (this_mpi_process == 0)
+    {
+      if (stat(output_dir_name.c_str(), &buffer) != 0)
+        {
+          create_output_folder(output_dir_name);
+        }
+    }
+
+  // Get the pointer of the only instance of the action manager
+  action_manager = DEMActionManager::get_action_manager();
+
+  // Set the simulation control as SimulationControlAdjointSteady
+  simulation_control = std::make_shared<SimulationControlSteady>(
+    parameters.simulation_control);
+
+  // Setup load balancing parameters and attach the correct functions to the
+  // signals inside the triangulation
+  // parameters.model_parameters.load_balance_method =
+  // Parameters::Lagrangian::ModelParameters<dim>::frequent;
+  load_balancing.set_parameters(parameters.model_parameters);
+
+  AdaptiveSparseContacts<dim, DEMProperties::PropertiesIndex>
```

Comment:
Hmm that's never fun, we might have to check if we can modify the load_balancing function to actually not have the requirement for a dummy object (for example passing nothing or having a default argument )